server/shared/storage/amazon.ts

```diff
@@ -0,0 +1,123 @@
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property
+  createReadStream(key: string): NodeJS.ReadableStream {
+    const params = { Bucket: this.#bucket, Key: key };
+    return this.#client.getObject(params).createReadStream();
+  }
+
+  // API docs: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property
+  createWriteStream(key: string): NodeJS.WritableStream {
+    const throughStream = miss.through();
+    const params = { Bucket: this.#bucket, Key: key, Body: throughStream };
+    this.#client.upload(params, () => null);
+    return throughStream;
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#putObject-property
+  async saveFile(key: string, data: Buffer): Promise<void> {
+    const params = { Bucket: this.#bucket, Key: key, Body: data };
+    await this.#client.putObject(params).promise();
+  }
+
+  // API docs: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#copyObject-property
+  async copyFile(key: string, newKey: string): Promise<void> {
+    const { base, ...rest } = path.parse(key);
+    const encodedSource = path.format({ base: encodeURIComponent(base), ...rest });
+    const params = {
+      Bucket: this.#bucket,
+      CopySource: this.path(`/${encodedSource}`),
+      Key: newKey
+    };
+    await this.#client.copyObject(params).promise();
+  }
+
+  async moveFile(key: string, newKey: string): Promise<void> {
+    await this.copyFile(key, newKey);
+    await this.deleteFile(key);
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObject-property
+  async deleteFile(key: string): Promise<DeleteResponse> {
+    const params = { Bucket: this.#bucket, Key: key };
+    await this.#client.deleteObject(params).promise();
+    return { isDeleted: true };
+  }
+
+  // API docs: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObjects-property
+  async deleteFiles(keys: string[]): Promise<DeleteResponse> {
+    const objects = keys.map(key => ({ Key: key }));
+    const params = {
+      Bucket: this.#bucket,
+      Delete: { Objects: objects }
+    };
+    await this.#client.deleteObjects(params).promise();
+    return { isDeleted: true };
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjects-property
+  async listFiles(key: string): Promise<FileListResponse[]> {
+    const params = { Bucket: this.#bucket, Prefix: key };
+    const { Contents: files } = await this.#client.listObjectsV2(params).promise();
+    return files.map(file => ({ path: file.Key }));
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#headObject-property
+  fileExists(key: string): Promise<boolean> {
+    const params = { Bucket: this.#bucket, Key: key };
+    return this.#client.headObject(params).promise()
+      .then(Boolean)
+      .catch(() => false);
```

Comment:
Should we match against `NotFound` and rest to enable specific error handling (i.e. if file exists but the access is forbidden)?